compiler-core/src/type_/environment.rs

```diff
@@ -587,4 +636,70 @@ impl<'a> Environment<'a> {
             .map(|t| t.to_string())
             .collect()
     }
+
+    /// Checks that the given patterns are exhaustive for given type.
+    /// Currently only performs exhaustiveness checking for custom types,
+    /// only at the top level (without recursing into constructor arguments).
+    pub fn exhaustive(
+        &mut self,
+        patterns: Vec<Pattern<PatternConstructor, Arc<Type>>>,
+        value_typ: Arc<Type>,
+    ) -> Option<Vec<String>> {
+        match &*value_typ {
+            Type::App {
+                name: type_name,
+                module: module_vec,
+                ..
+            } => {
+                let m = if module_vec.is_empty() || module_vec == self.current_module {
+                    None
+                } else {
+                    Some(module_vec.join("/"))
+                };
+                let mut matched_constructors = Vec::new();
```

Comment:
Let's use HashSets for the matched and unmatched constructors? They're unique so it fits a little better, and it means we can avoid the iteration through the matched constructors below when determining which are unmatched.

We could either initially construct the set with the constructors and remove them as we see them, or we could do something similar to the current approach and take the difference of the two sets.

Reply:
Done, I am not fluent with iterators yet (`.iter()`,`.into_iter()`, `.cloned()` etc) so this part may be worth double-checking.

Also because the HashSet doesn't keep a stable order, I had to do `.sorted()` when converting to a vec when returning unmatched constructors, to make the tests pass reliably

Reply:
This was done in https://github.com/gleam-lang/gleam/pull/1462/commits/c85818160140f801824ce5b1ff7b71468562c0b0